File: 0666-path-sum-iv/0666-path-sum-iv.cpp
class Solution {
public:
    unordered_map<int,int> tree;
    int sum = 0;
    int pathSum(vector<int>& nums) {
        for(int x : nums) 
            tree[x/10] = x%10;
        traverse(nums[0]/10, 0);
        return sum;
    }
    void traverse(int key, int preSum) {
        int level = key/10;
        int pos = key%10;
        int left = (level+1)*10 + pos*2-1; int right = (level+1)*10+pos*2;
        int curSum = preSum + tree[key];
        if(!tree.count(left) && !tree.count(right)) {
            sum += curSum;
            return;
        }
        if(tree.count(left)) traverse(left, curSum);
        if(tree.count(right)) traverse(right, curSum);
    }
};

File: 0666-path-sum-iv/0666-path-sum-iv.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var pathSum = function(nums) {
    let sum = 0;
    let tree = new Map();
    for(let x of nums) tree.set(Math.floor(x/10), x%10);

    function dfs(root, pre) {
        let cur = pre + tree.get(root);
        let lev = Math.floor(root/10); let pos = root%10;
        let left = (lev+1)*10 + 2*pos-1; let right = (lev+1)*10 + 2*pos;
        if(!tree.has(left) && !tree.has(right)) {
            sum += cur;
            return;
        }
        if(tree.has(left)) dfs(left, cur);
        if(tree.has(right)) dfs(right, cur);
    }

    dfs(Math.floor(nums[0]/10), 0);
    
    
    return sum;
};